lib/Dialect/FIRRTL/Transforms/Lint.cpp

```diff
@@ -83,18 +94,60 @@ class Linter {
 
     return success();
   }
+
+  LogicalResult checkXmr(XMRDerefOp op) {
+    // XMRs under layers are okay.
+    if (op->getParentOfType<LayerBlockOp>() ||
+        op->getParentOfType<sv::IfDefOp>())
+      return success();
+
+    // The XMR is not under a layer.  This module must never be instantiated in
+    // the design.  Intentionally do NOT use "effective" design as this could
+    // lead to false positives.
+    if (!instanceInfo.anyInstanceInDesign(fModule))
+      return success();
+
+    // If the op has a single user, that user is a connect, and the connect is
+    // to an instance which is marked `lowerToBind`, then this is a pattern for
+    // inlining the XMR into the bound instance site.  This pattern is used by
+    // Grand Central, but not elsewhere.
+    if (op.getResult().hasOneUse()) {
```

Comment:
Is single user restriction necessary? I think it's ok if all users are used by bound instances.

Reply:
It shouldn't be. Additionally, this made me realize that _no user_ is also fine as an unused expression will also not be emitted. I switched to an `all_of` check in the updated https://github.com/llvm/circt/pull/8668/commits/d8aa3b69582e705482dd94f1b4f8a69c060a1dd2. This also updated all the tests to make sure there was a user of the XMR and adds a test of no-user in "design" being okay.